### PR TITLE
Back/ar 43/en tant quutilisateur je souhaite pouvoir utiliser le service discord depuis le front

### PR DESCRIPTION
**Pull request by Vivelis**
* **Checklist before requesting a review**
- [x] The commits messages follows our guidelines
- [x] I have performed a self-review of my code
- [x] I have documented my code
- [ ] Do we need to implement unit tests?

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of changes does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [ ] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **Describe your changes**

Add the discord service to front and back

* **Issue ticket number and link**

[AR-43](https://bravmm.atlassian.net/browse/AR-43)

[AR-43]: https://bravmm.atlassian.net/browse/AR-43?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **New Features**
  - Added new "Timer" and "Discord" entries to the About section with descriptions for actions and reactions.
  - Introduced Discord authentication and token registration for users.
  - Implemented a service to refresh Discord access tokens automatically.
  - Integration of Discord messaging as a reaction to user-defined actions.
  - Dynamic population of services based on environment variables.

- **Enhancements**
  - Updated Spotify authentication to handle bad configurations and missing data more robustly.

- **Bug Fixes**
  - Adjusted data handling in the front-end `AREAFormTest` component to accommodate new data structure.

- **Documentation**
  - Updated developer documentation with instructions for adding new services to the database.

- **Style**
  - Improved code formatting in `AREAForm.tsx` for better readability.

### CODE REVIEW

File: front/app/Components/Modal.tsx [hunk 1]
```diff
@@ -121,7 +121,7 @@ const ModalUI: React.FC<{
     if (reactionDatas !== undefined) {
       setStep(Step.VALIDATE_OR_ADD_FORM);
     }
-  }, [reactionDatas]);
+  }, [reactionDatas, Step.VALIDATE_OR_ADD_FORM]);
```
[the remark this answers is not in the document]
Comment (Vivelis) [reply front/app/Components/Modal.tsx:124]: if I don't do that, it throw a warning

File: back/src/middleware/services/Auth/discordAuth.middleware.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    } catch (error) {
+        if (error instanceof Error) {
+            res.status(500).json({ error: error })
+            return;
+        }
+        console.error(error)
+        res.status(500).json({ error: 'Internal server error' })
+        return;
+    }
+}
```
[the remark this answers is not in the document]
Comment (Vivelis) [reply back/src/middleware/services/Auth/discordAuth.middleware.ts:51]: it is not mandatory

File: front/app/AREAFormTest/page.tsx [hunk 3]
```diff
@@ -13,7 +13,8 @@ const Page = () => {
         { title: "Date", type: "date" },
         // Add more fields as needed
     ];
-    const [datas, setDatas] = React.useState<{ [key: string]: string }>({});
+    type DatasType = { [key: string]: string }[];
+    const [datas, setDatas] = React.useState<string>("");
 
     useEffect(() => {
         if (Object.keys(datas).length !== 0) {
```
[the remark this answers is not in the document]
Comment (Vivelis) [reply front/app/AREAFormTest/page.tsx:20]: does using JSON.parse() could fix it ?

File: back/src/controller/services/discord.controller.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        if (!process.env.DISCORD_SERVICE_ID || Number(process.env.DISCORD_SERVICE_ID) !== serviceId) {
+            return res.status(400).json({ error: "Wrong service id" })
+        }
+        if (!accessToken || !refreshToken || !expiresIn) {
+            return res.status(401).json({ error: "No tokens provided" })
+        }
+        if (!req.params.guildId) {
+            return res.status(401).json({ error: "No guild ID provided" })
+        }
+        console.log("registerToken :\n accessToken : " + accessToken + "\n refreshToken : " + refreshToken + "\n expiresIn : " + expiresIn + "\n serviceId : " + serviceId + "\n userInfo : " + userInfo.userId);
```
[the remark this answers is not in the document]
Comment (Vivelis) [reply back/src/controller/services/discord.controller.ts]: I want to separate declaration and execution

File: back/src/controller/services/discord.controller.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ */
+const registerToken = async (req: Request, res: Response): Promise<Response> => {
+    const userInfo: TokenData = (req as CustomRequest).user
+
+    try {
+        const { accessToken, refreshToken, expiresIn, serviceId } = req.body
+        const encryptedAccessToken: { iv: string; content: string } = EncryptionService.encrypt(accessToken)
+        const encryptedRefreshToken: { iv: string; content: string } = EncryptionService.encrypt(refreshToken)
+
+        if (!process.env.DISCORD_SERVICE_ID || Number(process.env.DISCORD_SERVICE_ID) !== serviceId) {
```
[the remark this answers is not in the document]
Comment (Vivelis) [reply back/src/controller/services/discord.controller.ts:48]: I want the declaration and the calculation to be separated.

**Review by Beafowl-Pull — APPROVED** [review]
RAS

**Review by romainpanno — DISMISSED** [review]

**Review by KitetsuK — APPROVED** [review]

**Review by romainpanno — APPROVED** [review]

**Pull request merged**